beringei/tools/plain_text_service/FetchHandler.cpp

```diff
@@ -0,0 +1,151 @@
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#include "FetchHandler.h"
+
+#include <utility>
+
+#include "DateUtils.h"
+
+#include <folly/DynamicConverter.h>
+#include <folly/io/IOBuf.h>
+#include <proxygen/httpserver/ResponseBuilder.h>
+#include <thrift/lib/cpp/util/ThriftSerializer.h>
+#include <thrift/lib/cpp2/protocol/Serializer.h>
+#include <sstream>
+
+using apache::thrift::SimpleJSONSerializer;
+using namespace proxygen;
+
+namespace facebook {
+namespace gorilla {
+
+FetchHandler::FetchHandler(
+    std::shared_ptr<BeringeiConfigurationAdapterIf> configurationAdapter)
+    : RequestHandler(), configurationAdapter_(configurationAdapter) {}
+
+void FetchHandler::onRequest(std::unique_ptr<HTTPMessage> msg) noexcept {
+  _start = msg->getQueryParam("start");
+  _end = msg->getQueryParam("end");
+  _key = msg->getQueryParam("key");
+
+  LOG(INFO) << "Query start: " << _start << " end: " << _end
+            << " key: " << _key;
+}
+
+void FetchHandler::onBody(std::unique_ptr<folly::IOBuf> /* unused */) noexcept {
+  // nothing to do
+}
+
+void FetchHandler::onEOM() noexcept {
+  if (_start.empty()) {
+    return ResponseBuilder(downstream_)
+        .status(422, "Unprocessable Entity")
+        .header("Content-Type", "text/plain")
+        .body("start is required\r\n")
+        .sendWithEOM();
+  }
+
+  if (_end.empty()) {
+    return ResponseBuilder(downstream_)
+        .status(422, "Unprocessable Entity")
+        .header("Content-Type", "text/plain")
+        .body("end is required\r\n")
+        .sendWithEOM();
+  }
+
+  if (_key.empty()) {
+    return ResponseBuilder(downstream_)
+        .status(422, "Unprocessable Entity")
+        .header("Content-Type", "text/plain")
+        .body("key is required\r\n")
+        .sendWithEOM();
+  }
+
+  // parse times
+  time_t startTime;
+  if (!DateUtils::stringParseIsoTimestamp(_start, &startTime)) {
+    return ResponseBuilder(downstream_)
+        .status(422, "Unprocessable Entity")
+        .header("Content-Type", "text/plain")
+        .body("start is invalid\r\n")
+        .sendWithEOM();
+  }
+  time_t endTime;
+  if (!DateUtils::stringParseIsoTimestamp(_end, &endTime)) {
+    return ResponseBuilder(downstream_)
+        .status(422, "Unprocessable Entity")
+        .header("Content-Type", "text/plain")
+        .body("end is invalid\r\n")
+        .sendWithEOM();
+  }
+
+  // get client
+  BeringeiClient client(
+      configurationAdapter_, 1, BeringeiClient::kNoWriterThreads);
+
+  // build request
+  GetDataRequest request;
+  request.keys.emplace_back();
+  request.keys.back().key = _key;
+  request.keys.back().shardId = getShardId(_key, client.getNumShards());
+  request.begin = startTime;
+  request.end = endTime;
+
+  // get datapoints
+  gorilla::GorillaResultVector result;
+  client.get(request, result);
+
+  // setup res
```

Comment:
unnecessary + abbreviation (res).  For a comment here, "Transform datapoints into result." would work.